### PR TITLE
Check for (and reduce) overallocated stock when editing build orders.

### PR DESCRIPTION
**Pull request by mattbnz**
Check for (and reduce) overallocated stock when editing build orders.
    
After a build order is edited, check whether any bom items are now overallocated to the build, and if they are find appropriat build items to reduce the quantity of or remove from the build.

This feels like it could do with some tests... but I'm not yet familiar enough with Django's testing infrastructure to work out how to add the appropriate fixtures, etc for that sorry. 

Fixes: #3236

<a href="https://gitpod.io/#https://github.com/inventree/InvenTree/pull/3237"></a>

### CODE REVIEW

**Comment by SchrodingersGat** [comment]
@mattbnz as discussed in the other PR, I am not keen to have InvenTree silently adjust stock allocations without user input.

*However* I do very much like your implementation of the "trim" function, I think that should definitely stay. What we need is a method for the user to manually interact and instruct the allocation quantity to be trimmed, *if* the build is over allocated.

**Comment by mattbnz** [comment]
HI, I got pulled into another project at work and then school holidays hit so I haven't had a chance to get back to this in the last few weeks sorry.

It is still on my radar and todo list, to get back to ASAP, but realistically that's still a few weeks out, so if you want to forge ahead don't wait for me on anything, I'll see where things stand when I find the time to come back to this

Apologies.

**Comment by SchrodingersGat** [comment]
@mattbnz no worries at all. Just wanted to see where you were at! Ping me when you're ready to chat about this again.

**Comment by mattbnz** [comment]
Let me know what you think of this new approach.

**Comment by SchrodingersGat** [comment]
@mattbnz I like it :)

Only thing I think should be added is some additional unit tests for the 'build complete' endpoint which pass the various options through, and check that the correct operation is performed in each case.

File: InvenTree/build/serializers.py [hunk 1]
```diff
@@ -473,21 +473,36 @@ def save(self):
...
 
+class OverallocationChoice():
+    """Utility class to contain options for handling over allocated stock items."""
+
+    REJECT = 'reject'
+    ACCEPT = 'accept'
+    TRIM = 'trim'
+
+    OPTIONS = {
+        REJECT: ('not permitted'),
```
Comment (SchrodingersGat) [inline InvenTree/build/serializers.py:484]: Please capitalize these option strings

**Comment by SchrodingersGat** [comment]
@mattbnz trying this out on gitpod I am running into an error. I have over-allocated some stock and then try to complete as follows:

image

When selecting the "Deallocate" option the back-end throws an error for the "quantity" field - which is *not* displayed on the form and thus the user receives no feedback

**Comment by mattbnz** [comment]
Hmm, I'm not familiar with GitPod but I can look into it... is it doing anything special/weird? What testdata is it using?

I cannot reproduce locally atm fwiw.

**Comment by mattbnz** [comment]
I got gitpod running with the demo dataset and tested everything with no issues, same locally with the demo dataset (after guessing the admin password which I couldn't find documented anywhere...) 

Can you give some more details on how you reproduced/induced the error you're seeing please?

**Comment by SchrodingersGat** [comment]
> after guessing the admin password which I couldn't find documented anywhere..

https://inventree.readthedocs.io/en/latest/demo/

I'll update the `README.md` file for the demo dataset with these login details also.

**Comment by SchrodingersGat** [comment]
### Steps to Reproduce

#### Prepare Build Order

- Navigate to `/build/3/`
- Complete pending build output

image

#### Over Allocate Stock

Manually edit some allocations (via the table) so they are over-allocated:

image

#### Complete Build

- Reload page
- Select "Complete Build"
- Accept unallocated
- Select "Deallocate before completing.."

image

#### Submit Form

Get this error:

image

**Comment by mattbnz** [comment]
I've tried on both gitpod and locally, with the demo dataset and following those instructions and I cannot reproduce that behaviour sorry! 

I don't want to be presumptuous, but are you sure you don't have any other local changes in your client that are conflicting? The response shown in your screenshot doesn't make any sense, since the text of that error is only ever present in the "validate_accept_overallocated" method, nowhere near the quantity field, so it makes no sense/I can't see any viable path for how that text would even get associated with the quantity field during validation/serialization... 

Here's a [Screencast](https://user-images.githubusercontent.com/37299/183789832-eaa5b5b1-b4ef-427f-9b87-6f4130e40cac.webm) of what I see when running through those steps against the demo dataset against a freshly checked out copy of this branch. 

Sorry if I'm still missing something obvious...

**Comment by SchrodingersGat** [comment]
> I don't want to be presumptuous, but are you sure you don't have any other local changes in your client that are conflicting? 

That *may* have been it - after a full clean out of my browser cache it now works on the github.io instance.. Weird. It was not *running* on my machine but maybe there was something cached from my dev version which *was* running on my machine?

**Comment by SchrodingersGat** [comment]
@mattbnz I am happy to put it down to a cache gremlin on my end. Are you happy for me to merge this in now?

**Comment by mattbnz** [comment]
Cool, yeah, I'm happy with this - merge away, thanks!

**Comment by SchrodingersGat** [comment]
@mattbnz thanks for the contribution to the project, it is much appreciated! I think this feature is a very good addition.

If you are able, it would be great to be able to update the build documentation with this new option. I have created [a new issue](https://github.com/inventree/inventree-docs/issues/351) on the documentation repo. If you are able to tackle that at some point, that would be fantastic. It does not have to be super comprehensive :)

**Pull request merged**